docs: intg spec add blob derivation and preimage for integration spec (#1718)
Layr-Labs/eigenda: docs/spec/src/integration/spec/6-secure-integration.md

# Secure Integration

This page is meant to be read by eigenda and rollup developers who are writing a secure integration and need to understand the details. For users who want a high-level understanding of what a secure integration is, please visit our [secure integration overview](https://docs.eigenda.xyz/integrations-guides/rollup-guides/integrations-overview) page instead.

Unlike a trusted integration, a secure integration must handle the malicious data PUT on L1 Ethereum. A malicious batcher can post an invalid or malformed DA certificate (DA Cert) to stall the roll-up derivation pipeline. A malicious proposer can publish wrong L2 state roots. To overcome incorrect data posted on L1,
we rely on the GET path of the rollup to reliably filter out incorrect data posted by batcher, and challenge the incorrect state root by malicious proposer.

> We will first introduce the derivation procedures and then discuss two implementations - **eigenda-proxy** and **Hokulea**.

Reviewer: +1 - we should also be using terms defined in the glossary to make it more rollup agnostic and be better compliant with other integration spec terminologies

- Blob derivation: As the input transitions downward to the rollup payload at the bottom, the DA cert can be routed to one of the terminal states depending on if the DA certs is recent enough, if it is valid or if can be decoded correctly.
- Preimage oracle: Some procedures during the derivation requires fetching EigenDA blobs and other cert metadata. Access to each piece of data is abstracted behind a pre-image oracle. More see this [section](#eigenda-blob-derivation-and-preimage-oracle).
- Host: the name refers to an entity hostinh the preimage server behind the preimage oracle.

![](../../assets/integration/eigenda-blob-derivation.png)

### Terminal states
All inputs to EigenDA derivation pipeline ends in exactly one of four terminal states:
| State               | Meaning                                                             |
| ------------------- | ------------------------------------------------------------------- |
| **dropped**         | The input is rejected and therefore ignored by the rollup execution |
| **stalled**         | Required data is temporarily unavailable.                           |

> | **panic**           | A fundamental invariant (e.g., field-element validity) is violated. |

Author: What is your question here?

> | **panic**           | A fundamental invariant (e.g., field-element validity) is violated. |

Author: panic is removed

| **roll-up payload** | Success case when the desired payload bytes are produced.           |

### Failure cases

If an arrow in the diagram is labelled failed, the supplied data did not satisfy EigenDA’s safeguards. The DA Cert is discarded and nothing is forwarded downstream.

A path marked stall means the host cannot (yet) provide the requested data.
A path marked panic means the blob’s coefficients are not valid BN254 field elements. Then in theory, a KZG commitment cannot be generated, therefore derivation panic.

- parse failed – The batcher submitted an improperly-serialized DA Cert.

> - recency check failed – The DA Cert reached the roll-up inbox too late.

Author: I think it's implied.

> - recency check failed – The DA Cert reached the roll-up inbox too late.

Reviewer 2: It's not.

> - recency check failed – The DA Cert reached the roll-up inbox too late.

Author: I would just make it precise.

> - cert validity check failed – Either the certificate does not satisfy the [quorum-attestation constraint](#2-cert-validation) or the host lies about its validity via the preimage oracle.

Author: In Nitro M2 integration where cert valitidy is checked offchain, the preimage host can lie about the validity of the cert, @ethenotethan ?

> - cert validity check failed – Either the certificate does not satisfy the [quorum-attestation constraint](#2-cert-validation) or the host lies about its validity via the preimage oracle.

Reviewer: @bxue-l2 that is correct, with M2 we plan to do a prefix encoding of the cert's validity into the sequencer message bytes

> - cert validity check failed – Either the certificate does not satisfy the [quorum-attestation constraint](#2-cert-validation) or the host lies about its validity via the preimage oracle.

Author: Wait, does it imply that the sequencer posted something on L1 but claimed the posted thing is invalid?

- decode blob failed – The host supplies an EigenDA blob that cannot be decoded into back to rollup payload per the [spec](./3-datastructs.md#data-structs). This can occur when
  - the batcher intentionally corrupts the encoding, or
  - the host transmits incorrect data.

But if none of these errors arise, the derivation pipeline outputs the expected payload.

## EigenDA Blob derivation and Preimage Oracle

We have defined the canonical derivation process, and now we described two concrete implementations and their preimage oracle:

### EigenDA-proxy (Go)

Eigenda proxy is the first implementation of the eigenda blob derivation described above, it is also the first implementation  of the  preimage oracle

> server that uses the external RPC to implement the preimage oracle. The proxy also contains services like http server, secondary storage and more features on [proxy github page](https://github.com/Layr-Labs/eigenda-proxy?tab=readme-ov-file#features-and-configuration-options-flagsenv-vars). Below we highlight some parts around derivation pipeline

Reviewer 2: wdym by "the external RPC"?

Reviewer 2: doesn't belong here imo. Also "http server" is not "a service", its "the only service" that proxy provides

- the calldata input in proxy is received from http listening port (typically the requester is a roll-up consensus node).

> - proxy has both ETH RPC and EigenDA network RPC to get the preimage like the diagram below

Reviewer 2: "get the preimage". You never mentioned anywhere what the "preimage" means. Like its a preimage to what? the serialized cert. Typically preimage is preimage of a hash, so this is a bit of a stretch. So we should define it at the very beginning imo

> - Queries an Ethereum RPC endpoint to validate the cert. In practice, it is handled by implemented by `verification` Go package.

Reviewer 2: ```suggestion
- Queries an Ethereum RPC endpoint to validate the cert.
```

Reviewer 2: This should only be a spec imo. We should stay high-level. You could just mention in the beginning that the proxy is just a simple REST server that wraps our eigenda clients, and that all the logic is implemented in the clients, which could be used directly instead of via the proxy.

Author: The intent is yo highlight the fact that proxy relies on those rpc as preimage, do you get that part?

> - the EigenDA blob can be retrieved from EigenDA network either directly with an EigenDA relay node RPC or via distributed retrieval by communicating directly to the endpoints exposed by EigenDA validators. In practice, it is handled by implemented by `payloadretrieval` Go package.

Reviewer 2: ```suggestion
- the EigenDA blob can be retrieved from the EigenDA network either directly via an EigenDA relay node RPC or via distributed retrieval by communicating directly to the endpoints exposed by EigenDA validators.
```

- if everything succeeds, the roll-up payload is returned by the proxy to the requester via the HTTP response.

Eigenda-Proxy is the implemention which rollup consensus node runs in the normal mode for optimistic rollup.  

![](../../assets/integration/preimage-derivation-layers.png)

### Hokulea (Rust)

> We have developed another implementation of EigenDA blob derivation for OP stack, primarily for securing OP stack rollup with optimistic fault-proof,
> ZK Fault Proof or ZK proof. The hokulea client (and crates it uses) implements the identical eigenda blob derivation logics as mentioned in the [eigenda blob derivation section](#eigenda-blob-derivation), the client is intended to be imported as a library into OP consensus rust implementation [Kona](https://github.com/op-rs/kona).

Reviewer 2: ```suggestion
We have developed another implementation of EigenDA blob derivation for OP stack, which supports all of optimistic and ZK fault proof,
as well as validity ZK proofs. The hokulea client (and crates it uses) implement the eigenda blob derivation logic described in the [eigenda blob derivation section](#eigenda-blob-derivation). The client is intended to be imported as a library into OP consensus rust implementation [Kona](https://github.com/op-rs/kona).
```

The major difference between the two implementations comes to how the preimage oracle is abstracted and implemented.

> For hokulea, the interface is abstracted as key-value map to make preimage oracle verifiable on L1 Ethereum,
> whose relation is verifiable on L1 Ethereum.

Reviewer 2: ```suggestion
For Hokulea, the interface is abstracted as key-value map to make preimage oracle verifiable on L1 Ethereum.
```

Proxy must be able to return the encoded payload alone, though the specific implementation has not been decided yet.
As described in [data-struct page](./3-data-structs.md). During the creation of a DA cert a rollup payload, the data is first transformed into the encoded paylod
then undergo IFFT transform into the right polynomial format before sending to EigenDA that created a DA cert.

Normally the DecodeBlob functional interface on the GET path of the proxy does the FFT transformation on the received EigenDA blob to convert into the encoded
payload which is the original polynomial format.
To remove redundant work which the upper layer (Hokulea host) has to do (i.e. the FFT step), the proxy when used as a component for providing preimage, it returns
the encoded payload as opposed to the raw eigenda blob, which has the IFFT transformation, which is needed for optimistic challenge for opening points onchain.

> In the following sections, We provide the definition for each check in the diagram above. 

Reviewer 2: ```suggestion
In the following sections, we provide the definition for each check in the diagram above. 
```

# Secure Integration

This page is meant to be read by eigenda and rollup developers who are writing a secure integration and need to understand the details. For users who want a high-level understanding of what a secure integration is, please visit our [secure integration overview](https://docs.eigenda.xyz/integrations-guides/rollup-guides/integrations-overview) page instead.

Unlike a trusted integration, a secure integration must be able to handle any malicious data posted on Ethereum L1:
- A malicious batcher can post an invalid or malformed DA certificate (DA Cert) to stall the roll-up derivation pipeline. 
- A malicious proposer can publish wrong L2 state roots. 
To overcome incorrect data posted on L1,
we rely on the GET path of the rollup to reliably filter out incorrect data posted by malicious batcher(s), and challenge incorrect state roots posted by malicious proposer(s).

We will first introduce the derivation procedures and then discuss two implementations - **eigenda-proxy** and **Hokulea**.

## EigenDA blob derivation

> This section introduces a canonical procedure for deriving a rollup payload from a DA Cert in the GET path. The diagram below illustrates a step-by-step transition from a raw DA Cert byte string to the final roll-up payload:

Reviewer: ```suggestion
This section introduces a canonical procedure for deriving a rollup payload from a DA Cert in the GET path. The diagram below illustrates a step-by-step transition from a raw DA Cert byte string to the final rollup payload:
```

Reviewer 2: agree with this. I've never seen `roll-up` in two words like that

Author: WisprFlow likes -, Now everything went through a claude. This should not happen.

- Preimage oracle: Some procedures during the derivation require fetching EigenDA blobs and other cert metadata. Access to each piece of data is abstracted behind a pre-image oracle. More see this [section](#eigenda-blob-derivation-and-preimage-oracle).
- Host: entity hosting a preimage server which can reply to preimage oracle requests.

![](../../assets/integration/eigenda-blob-derivation.png)

### Terminal states
All inputs to EigenDA derivation pipeline end in exactly one of four terminal states:
| State               | Meaning                                                             |
| ------------------- | ------------------------------------------------------------------- |
| **dropped**         | The input is rejected and therefore ignored by the rollup execution |
| **stalled**         | Required data is temporarily unavailable.                           |
| **panic**           | A fundamental invariant (e.g., field-element validity) is violated. |

> | **roll-up payload** | Success case when the desired payload bytes are produced.           |

Reviewer: ```suggestion
| **rollup payload** | Success case when the desired payload bytes are produced.           |
```

### Failure cases

> If an arrow in the diagram is labelled failed, the supplied data did not satisfy EigenDA’s safeguards. The DA Cert is discarded and nothing is forwarded downstream.

Reviewer: https://www.notion.so/eigen-labs/M2-x-V2-Arbitrum-x-EigenDA-Design-1bf13c11c3e080ccb2d8d3a3fc15b780?source=copy_link#1ca13c11c3e080e68744cf577ac9ba73

  - the batcher intentionally corrupts the encoding, or
  - the host transmits incorrect data.

If none of these error cases arise, the derivation pipeline outputs the expected payload.

## EigenDA Blob derivation and Preimage Oracle

We have defined the canonical derivation process, and now we describe two concrete implementations and their preimage oracle:

### EigenDA-proxy (Go)

Eigenda proxy is the first implementation of the eigenda blob derivation described above, it is also the first implementation  of the  preimage oracle
server that uses the external RPC to implement the preimage oracle. Below we highlight some parts around derivation pipeline
- the calldata input in proxy is received from http listening port (typically the requester is a roll-up consensus node).
- proxy has both ETH RPC and EigenDA network RPC to get the preimage like the diagram below
- Queries an Ethereum RPC endpoint to validate the cert.
- the EigenDA blob can be retrieved from the EigenDA network either directly via an EigenDA relay node RPC or via distributed retrieval by communicating directly to the endpoints exposed by EigenDA validators.

> - if everything succeeds, the roll-up payload is returned by the proxy to the requester via the HTTP response.

Reviewer: is this worth enumerating? seems like duplication with existing retrieval lifecycle spec

Author: Removed

This section describes the canonical procedure for deriving a rollup payload from a DA Certificate. This derivation is integral to rollup consensus and must be implemented in both rollup nodes and secure integrations.

### Current Implementations

- **EigenDA Proxy**
- **OP EigenDA Secure Integration** ([Hokulea](https://github.com/Layr-Labs/hokulea/tree/master))

### Derivation Process

The diagram below shows the step-by-step transformation from input to final rollup payload:

**Key Components:**
- **Input:** Serialized DA Cert (as calldata) + block number of DA Cert inclusion
- **Blob Derivation:** Routes DA cert through validation to one of several terminal states
- **Preimage Oracle:** Interface for fetching additional data during derivation
  - Implementation varies by requirement (e.g., key-value mapping for optimistic fault proofs)
- **Host:** Entity that provides preimage oracle responses

![](../../assets/integration/eigenda-blob-derivation.png)

### Terminal States

All inputs to the EigenDA derivation pipeline end in exactly one of these states:

| State | Description |
|-------|-------------|
| **Dropped** | Input rejected and ignored by rollup execution |

> | **Stalled** | Required preimage data temporarily unavailable |

Reviewer 2: not necessarily temporary right?

| **Rollup Payload** | ✅ Success - desired payload bytes produced |

### Failure Cases

When validation fails, the DA Cert is discarded and nothing is forwarded downstream:

#### Parse Failed
- Batcher submitted improperly-serialized or unrecognized DA Cert

#### Recency Check Failed
- DA Cert reached rollup inbox after reference block number + recency window 
- Host provides false recency window size that leads to failure

#### Cert Validity Check Failed
- Certificate doesn't satisfy [quorum-attestation constraint](#2-cert-validation)

> - Host provides false validity information via preimage oracle

Reviewer 2: Actually what is this trying to say? That the host-provided validity boolean doesn't match the one provided from the onchain preimage oracle?

#### Decode Blob Failed
- EigenDA blob cannot be decoded back to rollup payload per [spec](./3-datastructs.md#data-structs)
- Causes:

>   - Batcher intentionally corrupts encoding
>   - Host transmits incorrect data

Reviewer 2: ```suggestion
  - Host or Batcher intentionally corrupts encoding
```

**Success:** If no failures occur, the pipeline outputs the expected payload.

## Secure Integration Framework

Rollup consensus must cover all aspects of EigenDA blob derivation. Designers have two degrees of freedom:

> 1. **Derivation Split:** Choose which parts are checked on-chain (pessimistically) vs off-chain (optimistically)

Author: oh, that is a good point. Yes, it is everything pessimistic.

2. **Proving VM Choice:** Select the on-chain proving virtual machine

Each integration can be tailored to fit specifi rollup protocol constraints.

### Splitting EigenDA Blob Derivation

Rollups can split the derivation pipeline between on-chain and off-chain verification:

**Examples:**

> - **Arbitrum with EigenDA V1:** All components through cert validity checked in rollup inbox

Reviewer 2: ```suggestion
- **Arbitrum with EigenDA V1:** All steps up to the cert validity are checked onchain in the rollup inbox
```

> - **OP Integration:** Entire EigenDA blob derivation takes place off-chain 

Reviewer 2: not necessarily if using op succinct with validity proofs that are checked pessimistically for every cert/blob pair.

Author: I guess what I really want to emphasize is what part is handled by the proving system, versus all parts are handled by the native L1 consensus by executing it.

Author: Maybe another way to make a distinction is the execution is taking place of change.

### VM for Optimistic Fault Proof

In order to secure parts from the EigenDA blob derivation taking place off-chain:

1. **Integration Required:** Blob derivation must be imported into L2 consensus as a library
2. **Compilation:** The library compiles to instructions replayable by on-chain VM
3. **Security:** Off-chain derivation secured by proof system

> 4. **Complete Coverage:** Combined on-chain and off-chain logics covers entire derivation

Reviewer 2: This is also true for zkvms right? Feel like this is good and should be moved up somewhere more prominent?

Author: Still think this is a good place because in the immediate section above I introduced the splitting the duration.

Author: But I'm considered daring to move it because it looks like it's true for all integrations.

**Preimage Oracle Requirement:** Both on-chain and off-chain implementations needed. 

![](../../assets/integration/secure-integration-model.png)

### ZKVM for ZK Fault Proof

> **Key Difference:** ZKVMs don't require preimage oracles since all preimage verification can be attached to EigenDA blob derivation usable as a library.

Reviewer 2: dont understand what this means

Author: We'll reorganize it.

**Hybrid Approach:** Secure integrations can theoretically combine both ZKVM and optimistic fault proof VMs.

## EigenDA Blob Derivation in EigenDA Proxy

> We have dedicated page for secure integrations, but let's review the **EigenDA Proxy** GET path implementation, which has been used in rollup consensus nodes since EigenDA integration began. Proxy also implements WRITE path, solely used by rollup batcher for rollup liveness.

Reviewer 2: add link to the page
